quantile/cm/stream.go

```diff
@@ -0,0 +1,278 @@
+)
+
+// stream represents a data stream
+type stream struct {
+	capacity   int        // stream capacity
+	eps        float64    // desired epsilon for errors
+	quantiles  []float64  // sorted target quantiles
+	samplePool SamplePool // pool of samples
+	floatsPool FloatsPool // pool of float64 slices
+
+	closed          bool       // whether the stream is closed
+	numValues       int64      // number of values
+	bufLess         minHeap    // sample buffer whose value is less than that at the insertion cursor
+	bufMore         minHeap    // sample buffer whose value is more than that at the insertion cursor
+	samples         sampleList // sample list
+	insertCursor    *Sample    // insertion cursor
+	compressCursor  *Sample    // compression cursor
+	compressMinRank int64      // compression min rank
+}
+
+// NewStream creates a new sample stream
+// TODO(xichen): the stream object itself should be pooled too
+// TODO(xichen): add metrics
+func NewStream(capacity int, opts Options) Stream {
+	samplePool := opts.SamplePool()
+	floatsPool := opts.FloatsPool()
+
+	return &stream{
+		capacity:   capacity,
+		eps:        opts.Eps(),
+		quantiles:  opts.Quantiles(),
+		bufLess:    minHeap(floatsPool.Get(capacity)),
+		bufMore:    minHeap(floatsPool.Get(capacity)),
+		samplePool: samplePool,
+		floatsPool: floatsPool,
+	}
+}
+
+func (s *stream) Add(value float64) {
+	s.addToBuffer(value)
+	s.insert()
+	s.compress()
+}
+
+func (s *stream) Flush() {
+	for s.bufLess.Len() > 0 || s.bufMore.Len() > 0 {
+		if s.bufMore.Len() == 0 {
+			s.resetInsertCursor()
+		}
+		s.insert()
+		s.compress()
+	}
+}
+
+func (s *stream) Quantile(q float64) float64 {
+	var (
+		minRank   int64
+		maxRank   int64
+		prev      = s.samples.Front()
+		curr      = s.samples.Front()
+		rank      = int64(math.Ceil(q * float64(s.numValues)))
+		threshold = int64(math.Ceil(float64(s.threshold(rank)) / 2.0))
+	)
+
+	for curr != nil {
+		maxRank = minRank + curr.numRanks + curr.delta
+		if maxRank > rank+threshold || minRank > rank {
+			break
+		}
+		minRank += curr.numRanks
+		prev = curr
+		curr = curr.next
+	}
+	if prev == nil {
+		return 0.0
+	}
+	return prev.value
+}
+
+func (s *stream) Reset() {
+	s.closed = false
+	s.numValues = 0
+	s.bufLess = minHeap(s.floatsPool.Get(s.capacity))
+	s.bufMore = minHeap(s.floatsPool.Get(s.capacity))
+	s.samples.Reset()
+	s.insertCursor = nil
+	s.compressCursor = nil
+	s.compressMinRank = 0
+}
+
+func (s *stream) Close() {
+	if s.closed {
+		return
+	}
+	s.closed = true
+	s.floatsPool.Put(s.bufLess)
+	s.floatsPool.Put(s.bufMore)
+	for sample := s.samples.Front(); sample != nil; sample = sample.next {
+		s.samplePool.Put(sample)
+	}
+}
+
+// addToBuffer adds a new sample to the buffer
+func (s *stream) addToBuffer(value float64) {
+	if s.numValues > 0 && value < s.insertPointValue() {
+		s.addToMinHeap(&s.bufLess, value)
+	} else {
+		s.addToMinHeap(&s.bufMore, value)
+	}
+}
+
+// insert inserts a sample into the stream
+func (s *stream) insert() {
```

Comment:
Before I read this I'm going to have to read the statsite implementation.  Might take a little time.